### DISCUSSION TITLE
How to use pfs from_dataframe

### DISCUSSION THREAD

**Discussion by Hendrik1987** [DHI/mikeio#519] [Q&A] [answered]
I have a pfs template with a dummy point output of one point and try to update this with a list of points from a dataframe `df2`.

The output in the template looks like this:
![image](https://user-images.githubusercontent.com/29595985/210337767-a9e9362d-20f3-4d2b-9083-b6253fdfbf6d.png)

I can update the number of points
![image](https://user-images.githubusercontent.com/29595985/210338221-07a0d10a-e540-4da7-9bc7-3e081fabb4b7.png)

and create pfs section of points from my dataframe:
![image](https://user-images.githubusercontent.com/29595985/210337863-2eb8a491-a098-42ba-a65f-dae6b1abc993.png)

However, I can't figure out how to replace the existing `OUTPUT_3.POINT_1` pfs section with the newly created `points` pfs section. 

I have tried:
1. `Out.OUTPUT_3 = points`, but then I am obviously loosing the header.
2. `Out.OUTPUT_3.POINT_1 = points`, but the I get a non-desired indent.
3. `Out.OUTPUT_3.POINT_1.update(points)` but it does not return anything.

Furthermore, I don't fully get 
- Why is `points` a `PfsSection`, albeit it is multiple sections on top of each other?
- On which `PfsSection` should I call the `.from_dataframe`? 

I have not managed to find the answers in the [docs](https://dhi.github.io/mikeio/pfs.html) or [example](https://github.com/DHI/mikeio/blob/main/notebooks/PFS.ipynb). Hence, thanks for any hints!

**Comment by jsmariegaard** [comment, marked as answer]
Your `points` object is a PfsSection of PfsSections. It behaves like a dictionary, hence you should be able to do something like this: 

    for n, p in points.items():
        Out.OUTPUT_3[n] = p

[.from_dataframe()](https://dhi.github.io/mikeio/pfs.html#mikeio.PfsSection.from_dataframe) is class method that returns a PfsSection. 

    sct = mikeio.PfsSection.from_dataframe(df)

Does that answer your questions?

**Reply by Hendrik1987** [reply]
> [.from_dataframe()](https://dhi.github.io/mikeio/pfs.html#mikeio.PfsSection.from_dataframe) is class method that returns a PfsSection.

Got it. So my code would probably be cleaner like this.

`points = mikeio.PfsSection.from_dataframe(df2, 'POINT_')`

> Your points object is a PfsSection of PfsSections. It behaves like a dictionary, hence you should be able to do something like this: [...]

Updated:

Yes this works:
```
for n, p in points.items():
    Out.OUTPUT_3[n] = p
```

Equally works to go via a dict:
```
dic = Out.OUTPUT_3.to_dict()
dic.update(points.to_dict())
Out.OUTPUT_3 = mikeio.pfs.PfsSection(dic)
```

The latter is a bit long, but from the documentation I did expect [.update](https://dhi.github.io/mikeio/pfs.html#mikeio.PfsSection.update) to do something similar. But this does not seem to work - maybe it's not intended?

```
Out.OUTPUT_3.update(points)
```

Also I had to remember to do 
```
df2.index = df2.index+1
```
The key is build from prefix and index [here](https://github.com/DHI/mikeio/blob/3380d4b7d2873bec2a1c5dcb05c487b165511ee2/mikeio/pfs.py#L479) and it seems that the key must be 1-based, otherwise the `.to_dataframe()` fails (and presumably the model run, too).

Finally, I wonder if a `.update_from_dataframe()` would be nice to wrap your solution above.

Thanks again!

**Comment by jsmariegaard** [comment]
I am not sure why you're trying to access OUTPUT_3[0]. OUTPUT_3 is a PfsSection dict-like structure and the keys (including subsections) are of str type. 

The intention with 

    for n, p in points.items():
        Out.OUTPUT_3[n] = p

was to get key-value pairs like this: 

    POINT_1, [PfsSection]
    POINT_2, [PfsSection]
    ...

But the code is could be wrong - I did not test.

**Reply by Hendrik1987** [reply]
No, that was me reading code suggestion too fast. `n` triggered for me that it is an `int` (like with `enumerate`), but of course it is the `key`, looping through `points`. Will give it a go.

**Reply by Hendrik1987** [reply]
Your solution works. Updated my comment above accordingly and accepted your answer. Thanks!
